src/zeit/content/volume/browser/form.py

```diff
@@ -87,23 +83,38 @@ def setUpWidgets(self, *args, **kw):
             self.widgets['volume'].setRenderedValue(settings.default_volume)
 
     def add(self, object):
-        path = self.volume_location(object)
-        # The last part of the path is the filename for the volume object.
-        volume_filename = path[-1]
-        container = zeit.cms.content.add.find_or_create_folder(*path[:-1])
-        if self._check_duplicate_volume(container, volume_filename):
+        folder, filename = self._create_folder(object, object.product.location)
+        if folder is None:
             return
-        container[volume_filename] = object
-        self._created_object = container[volume_filename]
-        self._finished_add = True
+        folder[filename] = object
+        self._created_object = folder[filename]
+
+        cp_template = zeit.cms.interfaces.ICMSContent(
+            object.product.cp_template, None)
+        if zeit.content.text.interfaces.IPythonScript.providedBy(cp_template):
+            folder, filename = self._create_folder(
+                object, object.product.centerpage)
+            if folder is None:
+                return
+            folder[filename] = cp_template(volume=object)
 
-    def volume_location(self, object):
-        location = object.fill_template(object.product.location)
-        location = location.replace(zeit.cms.interfaces.ID_NAMESPACE, '')
-        return [x for x in location.split('/') if x]
+        self._finished_add = True
 
-    def _check_duplicate_volume(self, location, name):
-        if location.get(name) is not None:
+    def _create_folder(self, volume, text):
+        path, filename = self._make_path(volume, text)
+        folder = zeit.cms.content.add.find_or_create_folder(*path)
+        if self._check_duplicate_item(folder, filename):
+            return (None, None)
+        return folder, filename
+
+    def _make_path(self, volume, text):
+        uniqueId = volume.fill_template(text)
+        uniqueId = uniqueId.replace(zeit.cms.interfaces.ID_NAMESPACE, '')
+        path = [x for x in uniqueId.split('/') if x]
+        return path[:-1], path[-1]
+
+    def _check_duplicate_item(self, folder, name):
+        if name in folder:
```

Comment:
Das funktioniert nicht zuverlässig, daher hatten wir `folder.get(name) is not None`. Problem ist, dass `name in folder` fälschlicherweise `False` sagt, wenn zwar ein Objekt des Namens da liegt, der Ordner im Cache aber veraltet ist (man also "Ordner neu laden" ausführen müsste). Bei `get` wird wirklich nachgeschlagen und der Inhalt neu gelesen. Weiß aber nicht ob das in Produktion ein Problem ist.